core/src/main/java/org/geysermc/geyser/entity/vehicle/VehicleComponent.java

```diff
@@ -0,0 +1,891 @@
+
+        float yaw = vehicle.getSession().getPlayerEntity().getYaw();
+        float sin = TrigMath.sin(yaw * TrigMath.DEG_TO_RAD);
+        float cos = TrigMath.cos(yaw * TrigMath.DEG_TO_RAD);
+        return Vector3f.from(input.getX() * cos - input.getY() * sin, 0, input.getY() * cos + input.getX() * sin);
+    }
+
+    protected Vector2f getVehicleRotation() {
+        LivingEntity player = vehicle.getSession().getPlayerEntity();
+        return Vector2f.from(player.getYaw(), player.getPitch() * 0.5f);
+    }
+
+    protected void moveVehicle(Vector3d javaPos) {
+        Vector3f bedrockPos = javaPos.toFloat();
+        Vector2f rotation = getVehicleRotation();
+
+        MoveEntityDeltaPacket moveEntityDeltaPacket = new MoveEntityDeltaPacket();
+        moveEntityDeltaPacket.setRuntimeEntityId(vehicle.getGeyserId());
+
+        if (vehicle.isOnGround()) {
+            moveEntityDeltaPacket.getFlags().add(MoveEntityDeltaPacket.Flag.ON_GROUND);
+        }
+
+        if (vehicle.getPosition().getX() != bedrockPos.getX()) {
+            moveEntityDeltaPacket.getFlags().add(MoveEntityDeltaPacket.Flag.HAS_X);
+            moveEntityDeltaPacket.setX(bedrockPos.getX());
+        }
+        if (vehicle.getPosition().getY() != bedrockPos.getY()) {
+            moveEntityDeltaPacket.getFlags().add(MoveEntityDeltaPacket.Flag.HAS_Y);
+            moveEntityDeltaPacket.setY(bedrockPos.getY());
+        }
+        if (vehicle.getPosition().getZ() != bedrockPos.getZ()) {
+            moveEntityDeltaPacket.getFlags().add(MoveEntityDeltaPacket.Flag.HAS_Z);
+            moveEntityDeltaPacket.setZ(bedrockPos.getZ());
+        }
+        vehicle.setPosition(bedrockPos);
+
+        if (vehicle.getYaw() != rotation.getX()) {
+            moveEntityDeltaPacket.getFlags().add(MoveEntityDeltaPacket.Flag.HAS_YAW);
+            moveEntityDeltaPacket.setYaw(rotation.getX());
+            vehicle.setYaw(rotation.getX());
+        }
+        if (vehicle.getPitch() != rotation.getY()) {
+            moveEntityDeltaPacket.getFlags().add(MoveEntityDeltaPacket.Flag.HAS_PITCH);
+            moveEntityDeltaPacket.setPitch(rotation.getY());
+            vehicle.setPitch(rotation.getY());
+        }
+        if (vehicle.getHeadYaw() != rotation.getX()) { // Same as yaw
+            moveEntityDeltaPacket.getFlags().add(MoveEntityDeltaPacket.Flag.HAS_HEAD_YAW);
+            moveEntityDeltaPacket.setHeadYaw(rotation.getX());
+            vehicle.setHeadYaw(rotation.getX());
+        }
+
+        if (!moveEntityDeltaPacket.getFlags().isEmpty()) {
+            vehicle.getSession().sendUpstreamPacket(moveEntityDeltaPacket);
+        }
+
+        ServerboundMoveVehiclePacket moveVehiclePacket = new ServerboundMoveVehiclePacket(javaPos.getX(), javaPos.getY(), javaPos.getZ(), rotation.getX(), rotation.getY());
+        vehicle.getSession().sendDownstreamPacket(moveVehiclePacket);
+        vehicle.getSession().setLastVehicleMoveTimestamp(System.currentTimeMillis());
```

Comment:
Not specifically a requirement now, but if we're tracking ticks we might be able to revisit some of these currentTimeMillis calls.